server/circleci.go

```diff
@@ -76,51 +75,63 @@ func (s *Server) triggerCircleCiIfNeeded(pr *model.PullRequest) {
 	mlog.Info("Triggered circleci", mlog.String("repo", pr.RepoName), mlog.Int("pr", pr.Number), mlog.String("fullname", pr.FullName))
 }
 
+func (s *Server) requestEETriggering(ctx context.Context, pr *model.PullRequest, eeBranch string, triggerBranch string) error {
+	circleciClient := http.DefaultClient
+
+	r, err := s.triggerEEPipeline(circleciClient, ctx, pr, eeBranch, triggerBranch)
+	if err != nil {
+		return err
+	}
+
+	workflowId, err := s.waitForWorkflowId(circleciClient, ctx, r.Id, s.Config.EnterpriseWorkflowName)
+	if err != nil {
+		return err
+	}
+
+	buildLink := "https://app.circleci.com/pipelines/github/" + s.Config.Org + "/" + s.Config.EnterpriseReponame + "/" + strconv.Itoa(r.Number) + "/workflows/" + workflowId
+	mlog.Debug("EE tests Workflow found", mlog.String("link", buildLink), mlog.String("wf id", workflowId), mlog.Int("pip number", r.Number))
+
+	err = s.waitForStatus(ctx, pr, s.Config.EnterpriseGithubStatusContext, "success")
+	if err != nil {
+		s.createEnterpriseTestsErrorStatus(ctx, pr, err)
+		return err
+	}
+
+	s.updateBuildStatus(ctx, pr, s.Config.EnterpriseGithubStatusEETests, buildLink)
+	return nil
+}
+
 type PipelineTriggeredResponse struct {
 	Number    int       `json:"number"`
 	State     string    `json:"state"`
 	Id        string    `json:"id"`
 	CreatedAt time.Time `json:"created_at"`
 }
 
-func (s *Server) triggerEnterprisePipeline(pr *model.PullRequest, eeBranch string, triggerBranch string) (string, error) {
+func (s *Server) triggerEEPipeline(circleciClient *http.Client, ctx context.Context, pr *model.PullRequest, eeBranch string, triggerBranch string) (*PipelineTriggeredResponse, error) {
 	body := strings.NewReader(`branch=` + eeBranch + `&parameters[external_branch]=` + triggerBranch + `&parameters[external_sha]=` + pr.Sha + `&parameters[external_pr]=` + strconv.Itoa(pr.Number))
-	req, err := http.NewRequest("POST", "https://circleci.com/api/v2/project/gh/"+s.Config.Org+"/"+s.Config.EnterpriseReponame+"/pipeline", body)
+	req, err := http.NewRequestWithContext(ctx, "POST", "https://circleci.com/api/v2/project/gh/"+s.Config.Org+"/"+s.Config.EnterpriseReponame+"/pipeline", body)
 	if err != nil {
-		return "", err
+		return nil, err
 	}
 	req.SetBasicAuth(os.ExpandEnv(s.Config.CircleCIToken), "")
 	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
-
-	client := http.Client{
-		Timeout: 20 * time.Second,
-	}
-	resp, err := client.Do(req)
+	resp, err := circleciClient.Do(req)
 	mlog.Debug("EE triggered", mlog.Int("pr", pr.Number), mlog.String("sha", pr.Sha), mlog.String("triggerRef", triggerBranch), mlog.String("eeBranch", eeBranch))
 	if err != nil {
-		return "", err
+		return nil, err
 	}
-	b, err := ioutil.ReadAll(resp.Body)
 	err = resp.Body.Close()
```

Comment:
We should not close the body before reading it. This should be done in a defer block.